### ISSUE TITLE
[Meta] dependency improvement: auto-dependencies via path-styled tree

### ISSUE THREAD

**Issue by Matze1224** [skonfig/skonfig#76]
As some already read in chat, I've asked why there wouldn't be auto-dependencies to the parent `__directory` object when you create a `__file` (see [#skonfig:matrix.org](https://matrix.to/#/#skonfig:matrix.org) at _1. Jan 2024_). So I wouldn't want to write any `require=` statements here but it should work without `CDIST_ORDER_DEPENDENCY`:

```sh
__directory /foo
__directory /foo/bar
__file /foo/bar/baz
```

This "auto-dependency" based on the object id can be generally implemented to make things easier. It shouldnt be restricted to only `__directory` and `__file`, as it can be utilised for more types using a path tree and thereby shouldn't be restricted to anything specific except flags via an environment variable. As backdraw, we lose backward compatibility to other manifests.

Manis pointed out in chat that the current dependency resolving isnt that easy with the current dependency resolving inherited by cdist. Copied his points here for easier read:

> Such a change to core is difficult to implement for different reasons. Firstly, we don't necessarily know for sure if __object_id is a file system path; it could work for __file but not for other types like __package. Also, such a change should only autorequire when a matching __directory object exists. This does not work with the require resolution in skonfig (as inherited from cdist) which resolves requires strongly (an error if no such object is created) and eagerly (runs an object as soon as the immediate requires are fulfilled which does not allow for globs in require; globs are not strictly required for this feature but go in the same direction).
tl;dr: the idea is great, but the feature is not easily implemented.

> In my mind I've been playing with other dependency resolution strategies for a while.
There is no easy solution to this problem because you want explorers and code-* to be run with as little as possible time in between while the manifest has to be run between the two possibly creating other objects changing the dependency graph.
And such an object created in the manifest could theoretically require=__the_last_object/in_manifest.
So what you'd want is to first create all objects, build the dependency graph and then walk through that.
But that doesn't work easily either because you need to run the objects' explorers in between.
In between because the manifest has access to their results.

> Maybe explorers should be split up into different categories
> - explorers which acquire information about the system and need to be run only once for all objects of a type,
> - explorers needed for general decisions in manifest
> - explorers which define what code gets generated.
>
> There are various cases for such autorequires, e.g. also in __key_value, __line, __block, __(postgres|mysql|...)_database
> The idea is really good, it's just not easy to patch into the current skonfig.
